### PR TITLE
Onboarding Refactor

### PR DESCRIPTION
**Pull request by aivantg**
[//]: # "These comments meant for your reference, they are invisible and don't need to be deleted"

## What's new in this PR?

This PR completely revamps the app's onboarding code to clean up some of the accumulated technical debt and work with the latest changes in the codebase.

**Major Changes:**
- Onboarding is now integrated with the latest schema updates (combined user tables and new field names)
- Onboarding takes advantage of new record transformations feature to minimize constant translating between airtable name and JS name
- Cleaner validation logic and new validator functions to more-systematically validate fields
- Integrated onboarding with proper routing (an onboarding user can log back in and will be returned to the onboarding step they were on, won't be able to access any other pages)
- Cleaned up airtable update flow to integrate seamlessly with redux and ensure that data is saved as user goes through onboarding
- Cleaned up Onboarding Render Logic by moving all onboarding data to it's own file
- Renamed all Onboarding sub components to have "Step" in the name and moved to a dedicated folder

### How to Review

Todo: Figure out the best way to review this so fang's life is easier 

### Related PRs

#69 Branches off of this PR

### Next steps

- Project Group Map View needs to be implemented
- Project Group Searching needs to be implemented
- Progress Bar Component needs to be fixed

### Tests Performed, Edge Cases

- Went through the onboarding flow multiple times, going back and forth and ensuring airtable stayed consistent
- Tested multiple validation edge cases along the way 

CC: @dfangshuo

### CODE REVIEW

File: src/lib/onboardingUtils.js [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+// Ensure value exists
+// Allows custom error message
+const validateExistence = (value, error = 'Required') => {
+  return value ? '' : error;
+};
+
+// Ensure valid and unique email
+const validateEmail = value => {
+  return /^[a-zA-Z0-9]+@[a-zA-Z0-9]+\.[A-Za-z]+$/.test(value)
```
Comment (dfangshuo) [inline src/lib/onboardingUtils.js]: this currently doesnt allow for emails in the form of {something}.{something}@email.com (when I tested). e.g. j.alex@gmail.com is flagged as an invalid email

File: src/components/AuthenticatedRoute.js [hunk 2]
```diff
@@ -2,35 +2,47 @@
...
+    if (isOnboarding(userCredentials)) {
+      return onboarding;
+    }
 
-    // If credential prop exists, ensure they are authorized
-    // If noauth prop exists, ensure they are NOT signed in
-    // else, just ensure they are signed in
     if (credential) {
+      // If credential prop exists, ensure they are authorized
+      // If noauth prop exists, ensure they are NOT signed in
```
[the remark this answers is not in the document]
Comment (aivantg) [reply src/components/AuthenticatedRoute.js]: sure sure

File: src/lib/adminUtils.js [hunk 3]
```diff
@@ -17,10 +17,13 @@ export async function removeOwner(owner) {
...
+  await refreshUserData(loggedInOwner.id);
 }
 
 export function getOwnerRecordsForProjectGroup(projectGroup) {
   const ownerPromises = projectGroup.ownerIds.map(getOwnerById);
-  return Promise.all(ownerPromises);
+  const allOwners = Promise.all(ownerPromises);
+
+  // Ensure onboarding users aren't considered
+  return allOwners.filter(o => o.onboardingStep === -1);
```
[the remark this answers is not in the document]
Comment (aivantg) [reply src/lib/adminUtils.js]: Yeah....not sure if there’s any way to get around it though if we want to somehow keep track of where users are during onboarding though. Realistically most people probably won’t leave mid-onboarding but it’s a nice small feature to have

File: src/lib/authUtils.js [hunk 4]
```diff
@@ -27,7 +27,7 @@ const loginUser = async (email, passwordHash) => {
 
     // Save key to redux store
     store.dispatch(authenticate(key));
-    refreshUserData(owner);
+    refreshUserData(owner.id);
```
[the remark this answers is not in the document]
Comment (aivantg) [reply src/lib/authUtils.js]: As in, the expected structure of the owner object? I use (and recommend we all use) `schema.js` as a guide to what columns exist on the object

File: src/lib/onboardingData.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
+import ContactInfoStep from '../screens/onboarding/steps/ContactInfoStep';
+import BasicInfoStep from '../screens/onboarding/steps/BasicInfoStep';
+import ProjectGroupStep from '../screens/onboarding/steps/ProjectGroupStep';
+import PaymentStep from '../screens/onboarding/steps/PaymentStep';
+import CompleteStep from '../screens/onboarding/steps/CompleteStep';
+import BylawStep from '../screens/onboarding/steps/BylawStep';
+import PaymentDetailsStep from '../screens/onboarding/steps/PaymentDetailsStep';
+
+// Maps Step to Component and Fields via Indices
+const OnboardingData = [
```
[the remark this answers is not in the document]
Comment (aivantg) [reply src/lib/onboardingData.js]: Haha yeah me neither

File: src/lib/paypalUtils.js [hunk 6]
```diff
@@ -62,7 +50,7 @@ const recordShareBuySuccess = async (details, data, values) => {
 
   // TODO: Couldn't this use the constructPaymentRecord function?
   const record = {
-    ownerId,
+    ownerId: [ownerId],
```
[the remark this answers is not in the document]
Comment (aivantg) [reply src/lib/paypalUtils.js]: Yeah, this is kinda awkward but for it’s linked records, airtable expects a list of ID’s, regardless of whether it’s a one-to-many or one-to-one relationship. I’m thinking potentially dealing with this in the schema generator at a low level, but it’s a little complicated so not there yet

File: src/lib/onboardingUtils.js (hunk 1 above)
Comment (aivantg) [reply src/lib/onboardingUtils.js]: Ooh great catch, I’ll add that

**Comment by aivantg** [comment]
The progress bar thing isn’t really a big, it’s just that in this branch I capitlized the file name and that broke things because git isn’t very sensitive to Capitlization. The solution is to manually rename the file to the capitalized version. I think when we start developing, it might make sense for everyone to check out a fresh copy of the repo so shouldn’t be a problem (once it’s merged in). 

Fixed the small email thing, let me know if there’s anything else!
